File: examples/python/matrix_multiply.py
# APG Example: matrix_multiply (Python/Numba)
# Computes C = A * B for NxN matrices stored as flat arrays.
# Expected parallelism: polyhedral (outer loop rows are independent)
import numpy as np

def matrix_multiply(A, B, C, N):
    for i in range(N):
        for j in range(N):
            tmp = 0.0
            for k in range(N):
                tmp = tmp + A[i * N + k] * B[k * N + j]
            C[i * N + j] = tmp

File: examples/rust/matrix_multiply.rs
// APG Example: matrix_multiply (Rust/Rayon)
// Computes C = A * B for NxN matrices stored as flat Vec<f32>.
// Expected parallelism: polyhedral (outer loop rows are independent)
pub fn matrix_multiply(a: &[f32], b: &[f32], c: &mut [f32], n: usize) {
    for i in 0..n {
        for j in 0..n {
            let mut tmp: f32 = 0.0;
            for k in 0..n {
                tmp += a[i * n + k] * b[k * n + j];
            }
            c[i * n + j] = tmp;
        }
    }
}
